lib/bolt/inventory/inventory2.rb

```diff
@@ -219,102 +198,199 @@ def expand_targets(targets)
           targets.split(/[[:space:],]+/).reject(&:empty?).map do |name|
             ts = resolve_name(name)
             ts.map do |t|
-              target = create_target(t)
-              target.inventory = self
-              target
+              # If the target exists, return it, otherwise create one
+              @targets[t] ? @targets[t]['self'] : create_target(t)
             end
           end
         end
       end
       private :expand_targets
 
-      def set_vars_from_hash(tname, data)
-        if data
-          # Instantiate empty vars hash in case no vars are defined
-          @target_vars[tname] ||= {}
-          # Assign target new merged vars hash
-          # This is essentially a copy-on-write to maintain the immutability of @target_vars
-          @target_vars[tname] = @target_vars[tname].merge(data).freeze
+      def ensure_in_all_group(targets)
+        names = flat_target_list(targets).map(&:name)
+        targets_to_add = Set.new(names) - @groups.target_names
+        targets_to_add.each { |t| add_to_group([@targets[t]['self']], 'all') }
+      end
+
+      def add_target(current_group, target, desired_group)
+        if current_group.name == desired_group
+          current_group.add_target(target.target_data_hash)
+          @groups.validate
+          update_target(target)
+          return true
+        end
+        # Recurse on children Groups if not desired_group
+        current_group.groups.each do |child_group|
+          add_target(child_group, target, desired_group)
         end
       end
-      private :set_vars_from_hash
+      private :add_target
+
+      # CODEREVEW: For now these methods live in inventory, should it be a subclass or some other module?
 
-      def set_facts(tname, hash)
-        if hash
-          @target_facts[tname] ||= {}
-          @target_facts[tname] = Bolt::Util.deep_merge(@target_facts[tname], hash).freeze
+      # This is effectively the init method for Target2
+      def create_target(target_name, target_hash = nil)
+        # Prefer target hash, then data from inventoryfile, allow for uri only with empty hash
+        data = target_hash || @groups.target_collect(target_name) || {}
+        data = { 'uri' => target_name } if data['uri'].nil? && data['name'].nil?
+        data['uri_obj'] = Bolt::Inventory::Inventory2.parse_uri(data['uri'])
+
+        if data['uri'] && data['name'].nil?
+          data['name'] = data['uri']
+          data['safe_name'] = data['uri_obj'].omit(:password).to_str.sub(%r{//}, '')
```

Comment:
Why do we remove the `//`?

Reply:
Is this for the case where we have a bare leading `//`? If so, then this should be `%r{^//}`

Reply:
Yes, it is to remove the `//` we append in `parse_uri` when we add it to initialized the addressable URI object with an empty scheme to ensure we parse the hostname correctly. I updated to only remove the `//` when it is the first two chars. Good catch.